README.md
# cfw

![GitHub](https://img.shields.io/github/license/bengarrett/cfw?style=flat)
[![Go Report Card](https://goreportcard.com/badge/github.com/bengarrett/cfw)](https://goreportcard.com/report/github.com/bengarrett/cfw) 
[![api](https://img.shields.io/static/v1?label=api&message=docs&color=<COLOR>&style=flat)](https://pkg.go.dev/github.com/bengarrett/cfw)
[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/bengarrett/cfw)](https://github.com/bengarrett/cfw/releases)

#### Package cfw contains ports of a few selected CFWheels v1 helpers that are used for string manipulation and have no Go equivalent.

>[CFWheels](https://cfwheels.org/) is an open-source [CFML (ColdFusion Markup Language)](http://lucee.org/) framework inspired by Ruby on Rails that provides fast application development, a great organization system for your code, and is just plain fun to use.

I created this package as I had been migrating CFWheels web applications over to Go. They had a few dependencies that need recreation to ensure a smooth transition.

☕ [Find cfw useful, why not buy me a coffee?](https://www.buymeacoffee.com/4rtEGvUIY)

## Documentation

https://pkg.go.dev/github.com/bengarrett/cfw

## Usages

```Go
package main

import "github.com/bengarrett/cfw"

func main() {

	// Obfuscate / DeObfuscate
	println(cfw.Obfuscate("1234"))
    println(cfw.DeObfuscate("a4363c"))
    
    // Excerpt
	println(cfw.Excerpt(
		"CFWheels: testing the excerpt view helper to see if it works or not.",
        "[more]", "excerpt view helper", 10))

    // Humanize
    println(cfw.Humanize("wheelsIsAFramework", nil))

    // Hyphenize
    println(cfw.Hyphenize("wheelsIsAFramework"))

    // StripLinks
    println(cfw.StripLinks(
        `Go to the <strong><a href="https://github.com/bengarrett/cfw">GitHub</a></strong> repo!`))

    // StripTags
    println(cfw.StripTags(
        `Go to the <strong><a href="https://github.com/bengarrett/cfw">GitHub</a></strong> repo!`))

    // TimeDistance
    add7Seconds := from.Add(time.Second * time.Duration(7)
    println(cfw.TimeDistance(time.Now(), add7Seconds), true))

    // Truncate
    println(cfw.Truncate(
        "CFW contains Go ports of a few selected CFWheels helpers.", "", 15))

    // WordTruncate
    println(cfw.WordTruncate(
        "CFW contains Go ports of a few selected CFWheels helpers.", "", 4))
    println(cfw.WordTruncate(
        "CFW contains Go ports of a few selected CFWheels helpers.", " 🥰", 3))
}
```

Will print

```bash
# Obfuscate / DeObfuscate
a4363c
1234
# Excerpt
[more]sting the excerpt view helper to see if[more]
# Humanize
Wheels Is A Framework
# Hyphenize
wheels-is-a-framework
# StripLinks
Go to the <strong>GitHub</strong> repo!
# StripTags
Go to the GitHub repo!
# TimeDistance
less than 10 seconds
# Truncate
CFW contains...
# WordTruncate
CFW contains Go ports...
CFW contains Go 🥰
```

##### Copyright © 2020 [Ben Garrett](mailto:code.by.ben@gmail.com), [MIT License](https://pkg.go.dev/github.com/fluhus/godoc-tricks?tab=licenses)

changelog.md
- `Humanize()` argument type change: `except []string`.
- Fixed some tests using incorrect argument names.

cfwheels_test.go
package cfw

import (
	"fmt"
	"testing"
	"time"
)

func ExampleDeObfuscate() {
	fmt.Println(DeObfuscate("9b1c6"))
	// Output: 1
}

func TestDeObfuscate(t *testing.T) {
	type args struct {
		s string
	}
	tests := []struct {
		name string
		s    string
		want string
	}{
		// values https://github.com/cfwheels/cfwheels/blob/1c3b9d6db79cdfbfbe49ae6816f6dc96262ccf82/wheels/tests/global/public/deobfuscateparam.cfc
		{"empty", "", ""},
		{"ok 1", "9b1c6", "1"},
		{"ok 2", "eb77359232", "999999999"},
		{"ok 3", "ac10a", "99"},
		{"ok 4", "b226582", "15765"},
		{"ok 5", "c06d44215", "69247541"},
		{"ok 6", "ac10a", "99"},
		{"error 1", "becca2515", "becca2515"},
		{"error 2", "a15ba9", "a15ba9"},
		{"error 3", "1111111111", "1111111111"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := DeObfuscate(tt.s); got != tt.want {
				t.Errorf("DeObfuscate() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleExcerpt() {
	pangram := "The quick brown fox jumps over the lazy dog"
	fmt.Println(Excerpt(pangram, "...", "The quick brown fox", 0))
	fmt.Println(Excerpt(pangram, "...", "", 19))
	fmt.Println(Excerpt(pangram, "🦊", "The quick brown ", 0))
	// Output: The quick brown fox...
	// The quick brown fox...
	// The quick brown 🦊
}

func TestExcerpt(t *testing.T) {
	type args struct {
		s       string
		replace string
		phrase  string
		n       int
	}
	const s = "CFWheels: testing the excerpt view helper to see if it works or not."
	tests := []struct {
		name string
		args args
		want string
	}{
		{"empty", args{}, "..."},
		{"1", args{s, "[more]", "CFWheels: testing the excerpt", 0}, "CFWheels: testing the excerpt[more]"},
		{"2", args{s, "[more]", "testing the excerpt", 0}, "[more]testing the excerpt[more]"},
		{"3", args{s, "[more]", "excerpt view helper", 10}, "[more]sting the excerpt view helper to see if[more]"},
		{"4", args{s, "[more]", "excerpt view helper", 25}, "CFWheels: testing the excerpt view helper to see if it works or no[more]"},
		{"5", args{s, "[more]", "see if it works", 25}, "[more]e excerpt view helper to see if it works or not."},
		{"6", args{s, "[more]", "jklsduiermobk", 25}, ""},
		{"utf8", args{"The quick brown 🦊 jumps over the lazy 🐕", "💬", "brown 🦊", 0}, "💬brown 🦊💬"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Excerpt(tt.args.s, tt.args.replace, tt.args.phrase, tt.args.n); got != tt.want {
				t.Errorf("Excerpt() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleHumanize() {
	fmt.Println(Humanize("helloWorldExample", nil))
	fmt.Println(Humanize("golangModOrVgo?", []string{"MOD", "VGO"}))
	// Output: Hello World Example
	// Golang MOD Or VGO?
}

func TestHumanize(t *testing.T) {
	type args struct {
		s      string
		except []string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		// values https://github.com/cfwheels/cfwheels/blob/1c3b9d6db79cdfbfbe49ae6816f6dc96262ccf82/wheels/tests/global/public/humanize.cfc
		{"empty", args{"", nil}, ""},
		{"lcase", args{"wheelsIsAFramework", nil}, "Wheels Is A Framework"},
		{"title", args{"WheelsIsAFramework", nil}, "Wheels Is A Framework"},
		{"ucase", args{"CFML", []string{"CFML"}}, "CFML"},
		{"except", args{"ACfmlFramework", []string{"CFML"}}, "A CFML Framework"},
		{"err 1", args{"wheelsIsACFMLFramework", nil}, "Wheels Is ACFML Framework"},
		{"same", args{"Some Input", nil}, "Some Input"},
		{"emoji", args{"theQuickBrown🦊JumpsOverTheLazy🐕", nil}, "The Quick Brown🦊 Jumps Over The Lazy🐕"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Humanize(tt.args.s, tt.args.except); got != tt.want {
				t.Errorf("Humanize() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleHyphenize() {
	fmt.Println(Hyphenize("aTourOfGo"))
	// Output: a-tour-of-go
}

func TestHyphenize(t *testing.T) {
	tests := []struct {
		name string
		s    string
		want string
	}{
		// values https://github.com/cfwheels/cfwheels/blob/13b9cb72f52a4edc7981ae572d050f8116af57ef/wheels/tests/global/strings.cfc
		{"empty", "", ""},
		{"ok 1", "wheelsIsAFramework", "wheels-is-a-framework"},
		{"ok 2", "WheelsIsAFramework", "wheels-is-a-framework"},
		{"ok 3", "aURLVariable", "a-url-variable"},
		{"ok 4", "URLVariable", "url-variable"},
		{"ucase", "ERRORMESSAGE", "errormessage"},
		{"lcase", "address", "address"},
		{"emoji", "TheQuickBrown🦊JumpsOverTheLazy🐕", "the-quick-brown🦊-jumps-over-the-lazy🐕"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Hyphenize(tt.s); got != tt.want {
				t.Errorf("Hyphenize() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleObfuscate() {
	fmt.Println(Obfuscate("1"))
	fmt.Println(Obfuscate("5551234"))
	// Output: 9b1c6
	// b3da865e
}

func TestObfuscate(t *testing.T) {
	tests := []struct {
		name string
		s    string
		want string
	}{
		// values https://github.com/cfwheels/cfwheels/blob/1c3b9d6db79cdfbfbe49ae6816f6dc96262ccf82/wheels/tests/global/public/obfuscateparam.cfc
		{"empty", "", ""},
		{"", "999999999", "eb77359232"},
		{"", "0162823571", "0162823571"},
		{"", "1", "9b1c6"},
		{"", "99", "ac10a"},
		{"", "15765", "b226582"},
		{"", "69247541", "c06d44215"},
		{"", "0413", "0413"},
		{"", "per", "per"},
		// in CFWheels this fails but in Go it returns a429646180a
		//{"", "1111111111", "1111111111"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Obfuscate(tt.s); got != tt.want {
				t.Errorf("Obfuscate() = %v, want %v", got, tt.want)
			}
		})
	}
}

func Test_reverseInt(t *testing.T) {
	tests := []struct {
		name    string
		i       int
		want    int
		wantErr bool
	}{
		{"zero", 0, 0, false},
		{"ok", 2345678, 8765432, false},
		{"505", 505, 505, false},
		{"leading 0", 0005, 5, false}, // expected behaviour?
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := reverseInt(tt.i)
			if (err != nil) != tt.wantErr {
				t.Errorf("reverseInt() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if got != tt.want {
				t.Errorf("reverseInt() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleStripLinks() {
	fmt.Println(StripLinks(`<a href="https://golang.org">The Go Programming Language</a>.`))
	// Output: The Go Programming Language.
}

func TestStripLinks(t *testing.T) {
	// values: https://github.com/cfwheels/cfwheels/blob/9eea7a6ac77956c8825e037f2e3e6b8c0f346267/wheels/tests/view/sanitize/striptags.cfc
	str := `this <a href="http://www.google.com" title="google">is</a> a <a href="mailto:someone@example.com" title="invalid email">test</a> to <a name="anchortag">see</a> if this works or not.`
	emoji := `The quick <b><a href="https://example.com">brown 🦊</a></b> jumps over the lazy 🐕`
	tests := []struct {
		name string
		s    string
		want string
	}{
		{"empty", "", ""},
		{"string", str, "this is a test to see if this works or not."},
		{"emoji", emoji, "The quick <b>brown 🦊</b> jumps over the lazy 🐕"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := StripLinks(tt.s); got != tt.want {
				t.Errorf("StripLinks() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleStripTags() {
	fmt.Println(StripTags(`<h1><a href="https://golang.org">The Go Programming Language</a>.</h1>`))
	// Output: The Go Programming Language.
}

func TestStripTags(t *testing.T) {
	// values: // https://github.com/cfwheels/cfwheels/blob/1c3b9d6db79cdfbfbe49ae6816f6dc96262ccf82/wheels/tests/view/text/excerpt.cfc
	str := `<h1>this</h1><p><a href="http://www.google.com" title="google">is</a></p><p>a <a href="mailto:someone@example.com" title="invalid email">test</a> to<br><a name="anchortag">see</a> if this works or not.</p>`
	emoji := `The quick <b><a href="https://example.com">brown 🦊</a></b> jumps over the lazy 🐕`
	tests := []struct {
		name string
		s    string
		want string
	}{
		{"empty", "", ""},
		{"string", str, "thisisa test tosee if this works or not."},
		{"emoji", emoji, "The quick brown 🦊 jumps over the lazy 🐕"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := StripTags(tt.s); got != tt.want {
				t.Errorf("StripTags() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleTimeDistance() {
	const layout = "2006 Jan 2"
	f, _ := time.Parse(layout, "2000 Jan 1")
	t, _ := time.Parse(layout, "2020 Jun 30")
	fmt.Println(TimeDistance(f, t, false))
	// Output: over 20 years
}

func ExampleTimeDistance_seconds() {
	f, _ := time.Parse(time.RFC3339, "2006-01-02T15:04:05Z07:00")
	t, _ := time.Parse(time.RFC3339, "2006-01-02T15:04:10Z07:00")
	fmt.Println(TimeDistance(f, t, true))
	// Output: less than 5 seconds
}

func TestTimeDistance(t *testing.T) {
	type args struct {
		from time.Time
		to   time.Time
		sec  bool
	}
	var n = time.Now()
	tests := []struct {
		name string
		args args
		want string
	}{
		{"empty", args{}, "less than a minute"},
		{"<1m", args{n, n.Add(time.Second * time.Duration(5-1)), false}, "less than a minute"},
		{"<10s", args{n, n.Add(time.Second * time.Duration(10-1)), true}, "less than 10 seconds"},
		{"<20s", args{n, n.Add(time.Second * time.Duration(20-1)), true}, "less than 20 seconds"},
		{"1/2min", args{n, n.Add(time.Second * time.Duration(40-1)), true}, "half a minute"},
		{"<1min", args{n, n.Add(time.Second * time.Duration(59)), false}, "less than a minute"},
		{"1min", args{n, n.Add(time.Second * time.Duration(60+50)), false}, "1 minute"},
		{"44min", args{n, n.Add(time.Minute * time.Duration(45-1)), false}, "44 minutes"},
		{"1h", args{n, n.Add(time.Minute * time.Duration(90-1)), false}, "about 1 hour"},
		{"23h", args{n, n.Add(time.Minute * time.Duration(1440-1)), false}, "about 23 hours"},
		{"1d", args{n, n.Add(time.Minute * time.Duration(2880-1)), false}, "1 day"},
		{"29d", args{n, n.Add(time.Minute * time.Duration(43200-1)), false}, "29 days"},
		{"1m", args{n, n.Add(time.Minute * time.Duration(86400-1)), false}, "about 1 month"},
		{"11m", args{n, n.Add(time.Minute * time.Duration(525600-1)), false}, "11 months"},
		{"1y", args{n, n.Add(time.Minute * time.Duration(657000-1)), false}, "about 1 year"},
		{">1y", args{n, n.Add(time.Minute * time.Duration(919800-1)), false}, "over 1 year"},
		{"2y", args{n, n.Add(time.Minute * time.Duration(1051200-1)), false}, "almost 2 years"},
		{">2y", args{n, n.Add(time.Minute * time.Duration(1051200)), false}, "over 2 years"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := TimeDistance(tt.args.from, tt.args.to, tt.args.sec); got != tt.want {
				t.Errorf("TimeDistance() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleTruncate() {
	fmt.Println(Truncate("Go is an open source programming language", "", 8))
	fmt.Println(Truncate("Go is an open source programming language", "?", 6))
	// Output: Go is...
	// Go is?
}

func TestTruncate(t *testing.T) {
	type args struct {
		s       string
		replace string
		n       int
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		// values: https://github.com/cfwheels/cfwheels/blob/9738bd71e7b8632587cd09a0a2d7d3e779700a7a/wheels/tests/view/text/truncate.cfc
		{"empty", args{}, ""},
		{"ok1", args{"this is a test to see if this works or not.", "[more]", 20}, "this is a test[more]"},
		{"err1", args{"", "[more]", 20}, ""},
		{"ok2", args{"this is a test to see if this works or not.", "", 20}, "this is a test to..."},
		{"emoji", args{"The quick brown 🦊 jumps over the lazy 🐕", "💬", 21}, "The quick brown 🦊💬"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Truncate(tt.args.s, tt.args.replace, tt.args.n); got != tt.want {
				t.Errorf("Truncate() = %v, want %v", got, tt.want)
			}
		})
	}
}

func ExampleWordTruncate() {
	fmt.Println(WordTruncate("Go is an open source programming language", "", 2))
	fmt.Println(WordTruncate("Go is an open source programming language", "?", 2))
	// Output: Go is...
	// Go is?
}

func TestWordTruncate(t *testing.T) {
	type args struct {
		s       string
		replace string
		n       int
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		// values: https://github.com/cfwheels/cfwheels/blob/9738bd71e7b8632587cd09a0a2d7d3e779700a7a/wheels/tests/view/text/wordtruncate.cfc
		{"empty", args{}, ""},
		{"ok", args{"CFWheels is a framework for ColdFusion", "", 4}, "CFWheels is a framework..."},
		{"emoji", args{"The quick brown 🦊 jumps over the lazy 🐕", "💬", 4}, "The quick brown 🦊💬"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := WordTruncate(tt.args.s, tt.args.replace, tt.args.n); got != tt.want {
				t.Errorf("WordTruncate() = %v, want %v", got, tt.want)
			}
		})
	}
}

// Test unique examples used in README.md
func TestReadmeExamples(t *testing.T) {
	var a, e, s string
	s = `Go to the <strong><a href="https://github.com/bengarrett/cfw">GitHub</a></strong> repo!`
	a = StripLinks(s)
	e = "Go to the <strong>GitHub</strong> repo!"
	if a != e {
		t.Errorf("mismatch, got: %v, want: %v", a, e)
	}
	a = StripTags(s)
	e = "Go to the GitHub repo!"
	if a != e {
		t.Errorf("mismatch, got: %v, want: %v", a, e)
	}
	from := time.Now()
	to := from.Add(time.Second * time.Duration(7)) // add 7 seconds
	a = TimeDistance(from, to, true)
	e = "less than 10 seconds"
	if a != e {
		t.Errorf("mismatch, got: %v, want: %v", a, e)
	}
	s = `CFW contains Go ports of a few selected CFWheels helpers.`
	a = Truncate(s, "", 15)
	e = "CFW contains..."
	if a != e {
		t.Errorf("mismatch, got: %v, want: %v", a, e)
	}
	s = `CFW contains Go ports of a few selected CFWheels helpers.`
	a = WordTruncate(s, "", 4)
	e = "CFW contains Go ports..."
	if a != e {
		t.Errorf("mismatch, got: %v, want: %v", a, e)
	}
}


cfwheels.go
// Package cfw contains ports of a few selected CFWheels helpers that
// are used for string manipulation and have no Go equivalent.
// © Ben Garrett https://github.com/bengarrett/cfw
package cfw

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

// DeObfuscate de-obfuscates a CFWheels obfuscateParam or
//  Obfuscate()
// obfuscated string.
func DeObfuscate(s string) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	if _, err := strconv.Atoi(s); err == nil || len(s) < 2 {
		return s
	}
	// De-obfuscate string.
	new := s[2:] // last 2 chars
	zi, err := strconv.ParseInt(new, 16, 0)
	if err != nil {
		return s
	}
	zi = zi ^ 461 // bitxor
	zs := strconv.Itoa(int(zi))
	l := len(zs) - 1
	new = ""
	for i := 0; i < l; i++ {
		f := zs[l-i:][:1]
		new += f
	}
	// Create checks.
	ct := 0
	l = len(new)
	for i := 0; i < l; i++ {
		chr := new[i : i+1]
		rvi, err := strconv.Atoi(chr)
		if err != nil {
			return s
		}
		ct = ct + rvi
	}
	// Run checks.
	ci, err := strconv.ParseInt(s[:2], 16, 0)
	if err != nil {
		return s
	}
	c2 := strconv.FormatInt(int64(ci), 10)
	if strconv.FormatInt(int64(ct+154), 10) != c2 {
		return s
	}
	return new
}

// Excerpt replaces n characters from s which match the first instance of a given phrase.
func Excerpt(s, replace, phrase string, n int) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	if replace == "" {
		replace = "..."
	}
	pos := strings.Index(s, phrase)
	if pos < 0 {
		return ""
	}
	// Set start info based on whether the excerpt text found, including its radius, comes before the start of the string.
	sp, ts := 0, ""
	if (pos - n) > 1 {
		sp = pos - n
		ts = replace
	}
	// Set end info based on whether the excerpt text found, including its radius, comes after the end of the string.
	ep, te := len(s), ""
	if (pos + len(phrase) + n) <= len(s) {
		ep = pos + n
		te = replace
	}
	var mid string
	if ln := ep + len(phrase); ln >= len(s) {
		mid = s[sp:]
	} else {
		mid = s[sp:ln]
	}
	return ts + mid + te
}

// Humanize returns readable text by separating camelCase strings to multiple, capitalized words.
func Humanize(s string, except []string) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/632ea90547da368cddd77cefe17f42a7eda871e0/wheels/global/util.cfm
	// Add a space before every capitalized word.
	s = regexp.MustCompile(`([A-Z])`).ReplaceAllString(s, " $1")
	// Fix abbreviations so they form a word again (example: aURLVariable).
	s = regexp.MustCompile(`([A-Z])\s([A-Z])(?:\s|\b)`).ReplaceAllString(s, "$1$2")
	// Handle exceptions.
	for _, e := range except {
		// (?i) case-insensitive
		s = regexp.MustCompile(`(?i)`+e+`(?:\b)`).ReplaceAllString(s, e)
	}
	// Support multiple word input by stripping out all double spaces created.
	s = regexp.MustCompile(`(\s\s)`).ReplaceAllString(s, " ")
	// Capitalize the first letter and trim final result.
	s = strings.TrimPrefix(s, " ")
	return strings.Title(s)
}

// Hyphenize converts camelCase strings to a lowercase hyphened string.
func Hyphenize(s string) string {
	s = regexp.MustCompile(`([A-Z][a-z])`).ReplaceAllString(s, strings.ToLower(`-$1`))
	s = regexp.MustCompile(`([a-z])([A-Z])`).ReplaceAllString(s, strings.ToLower(`$1-$2`))
	s = regexp.MustCompile(`^-`).ReplaceAllString(strings.ToLower(s), "")
	return s
}

// Obfuscate a numeric string to insecurely hide database primary key values when passed along a URL.
func Obfuscate(s string) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	// Make sure string doesn't start with a zero.
	if len(s) <= 0 || s[0] == '0' {
		return s
	}
	atoi, err := strconv.Atoi(s)
	if err != nil {
		return s
	}
	// Count the number of digits in s.
	count := len(s)
	ri, err := reverseInt(atoi)
	if err != nil {
		return s
	}
	f64 := math.Pow10(count) + float64(ri)
	// Keep a and b as int types.
	a, b := int(f64), 0
	for i := 1; i <= count; i++ {
		// Slice individual digits from s and sum the results.
		ps, err := strconv.Atoi(string(s[count-i]))
		if err != nil {
			return s
		}
		b += ps
	}
	// Base64 conversion.
	a ^= 461
	b += 154
	return strconv.FormatInt(int64(b), 16) + strconv.FormatInt(int64(a), 16)
}

// StripLinks removes all HTML links from a string leaving just the link text.
func StripLinks(s string) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/daa7c43fc993cab00f52cf8ac881e6cc93c02fe1/wheels/view/sanitize.cfm
	return regexp.MustCompile(`<a.*?>(.*?)</a>`).ReplaceAllString(s, "$1")
}

// StripTags removes all HTML tags from a string.
func StripTags(s string) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/daa7c43fc993cab00f52cf8ac881e6cc93c02fe1/wheels/view/sanitize.cfm
	s = regexp.MustCompile(`<\ *[a-z].*?>`).ReplaceAllString(s, "")
	s = regexp.MustCompile(`<\ */\ *[a-z].*?>`).ReplaceAllString(s, "")
	return s
}

// TimeDistance describes the difference between two time values.
func TimeDistance(from time.Time, to time.Time, seconds bool) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	delta := to.Sub(from)
	s, m, h := int(delta.Seconds()), int(delta.Minutes()), int(delta.Hours())
	var days, months, years int
	switch {
	case m <= 1:
		if seconds != true {
			switch {
			case s < 60:
				return "less than a minute"
			default:
				return "1 minute"
			}
		}
		switch {
		case s < 5:
			return "less than 5 seconds"
		case s < 10:
			return "less than 10 seconds"
		case s < 20:
			return "less than 20 seconds"
		case s < 40:
			return "half a minute"
		default:
			return "1 minute"
		}
	case m < 45:
		return fmt.Sprintf("%d minutes", m)
	case m < 90:
		return "about 1 hour"
	case m < 1440:
		return fmt.Sprintf("about %d hours", h)
	case m < 2880:
		return "1 day"
	case m < 43200:
		days = h / 24
		return fmt.Sprintf("%d days", days)
	case m < 86400:
		return "about 1 month"
	case m < 525600:
		months = h / 730
		return fmt.Sprintf("%d months", months)
	case m < 657000:
		return "about 1 year"
	case m < 919800:
		return "over 1 year"
	case m < 1051200:
		return "almost 2 years"
	default:
		years = m / 525600
		return fmt.Sprintf("over %d years", years)
	}
}

// Truncate a string to the specified number and replace the trailing characters.
func Truncate(s, replace string, n int) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	if replace == "" {
		replace = "..."
	}
	if utf8.RuneCountInString(s) <= n {
		return s
	}
	return s[0:n-utf8.RuneCountInString(replace)] + replace
}

// WordTruncate truncates a string to the specified number of words and replaces the trailing characters.
func WordTruncate(s, replace string, n int) string {
	// CFML source: https://github.com/cfwheels/cfwheels/blob/cf8e6da4b9a216b642862e7205345dd5fca34b54/wheels/global/misc.cfm
	if replace == "" {
		replace = "..."
	}
	words := strings.Fields(s)
	if len(words) >= utf8.RuneCountInString(s) {
		return s
	}
	var str string
	for i, w := range words {
		if i+1 >= n {
			str += w
			break
		}
		str += w + " "
	}
	return str + replace
}

// reverseInt reverses an integer.
func reverseInt(i int) (reverse int, err error) {
	// credit: Wade73
	// http://stackoverflow.com/questions/35972561/reverse-int-golang
	itoa, str := strconv.Itoa(i), ""
	for x := len(itoa); x > 0; x-- {
		str += string(itoa[x-1])
	}
	reverse, err = strconv.Atoi(str)
	if err != nil {
		return reverse, err
	}
	return reverse, err
}
